Add doc for alerting comments (#7360)
opensearch-project/documentation-website: _observing-your-data/alerting/api.md

Use the Alerting API to programmatically create, update, and manage monitors and alerts. For APIs that support the composite monitor specifically, see [Managing composite monitors with the API]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/composite-monitors/#managing-composite-monitors-with-the-api). 

## Create a query-level monitor

> Query-level monitors run the query and check whether or not the results should trigger an alert. Query-level monitors can only trigger one alert at a time. For more information about query-level and bucket-level monitors, see [Creating monitors]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/).

Reviewer: ```suggestion
Query-level monitors run the query and determine whether or not the results should trigger an alert. Query-level monitors can only trigger one alert at a time. For more information about query-level and bucket-level monitors, see [Creating monitors]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/).
```

  }]
}
```
{% include copy-curl.html %}

> If you use a custom webhook for your destination and need to embed JSON in the message body, be sure to escape your quotes.

Reviewer: ```suggestion
If you use a custom webhook for your destination and need to embed JSON in the message body, be sure to escape the quotation marks:
```

  }
}
```
{% include copy-curl.html %}

</details>

> To specify a time zone, you can do so by including a [cron expression]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/cron/) with a time zone name in the `schedule` section of your request. The following example creates a monitor that runs at 12:10 PM Pacific Time on the 1st day of every month.

Reviewer: ```suggestion
To specify a time zone, you can do so by including a [cron expression]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/cron/) with a time zone name in the `schedule` section of your request. The following example creates a monitor that runs at 12:10 PM Pacific Time on the first day of each month.
```

```json
GET /_plugins/_alerting/findings/_search?
```
{% include copy-curl.html %}

To retrieve metadata for an individual document finding entry, you can search for the finding by its `findingId` as follows:

```json
GET /_plugins/_alerting/findings/_search?findingId=gKQhj8WJit3BxjGfiOXC
```
{% include copy-curl.html %}

The response returns the number of individual finding entries in the `total_findings` field.

> To get more specific results in a findings search, you can use any of the optional path parameters that are defined in the following table.

Reviewer: ```suggestion
To get more specific results in a findings search, you can use any of the optional path parameters defined in the following table.
```

You can create a document-level monitor with a POST request that provides the monitor details in the request body.
At a minimum, you need to provide the following details: specify the queries or combinations by tag with the `inputs` field, a valid trigger condition, and provide the notification message in the `action` field.

> The following table shows the syntax to use for each trigger option.

Reviewer: ```suggestion
The following table provides the syntax to use for each trigger option.
```

}

```
{% include copy-curl.html %}

### Limitations

If you run a document-level query while the index is getting reindexed, the API response will not return the reindexed results. To get updates, wait until the reindexing process completes, then rerun the query.

---

## Update monitor

> When updating a monitor, you can optionally include `seq_no` and `primary_term` as URL parameters. If these numbers do not match the existing monitor or the monitor doesn't exist, the Alerting plugin throws an error. OpenSearch increments the version number and the sequence number automatically (see the example response).

Reviewer 2: ```suggestion
When updating a monitor, you can optionally include `seq_no` and `primary_term` as URL parameters. If these numbers do not match the existing monitor or the monitor does not exist, the Alerting plugin throws an error. OpenSearch increments the version number and the sequence number automatically (see the example response).
```

  }
}
```
{% include copy-curl.html %}

</details>

---

## GET monitor

> #### Example request

Reviewer: Stacked headings

  }
}
```
{% include copy-curl.html %}

</details>

---

## DELETE monitor

> #### Example request

Reviewer: Stacked headings

Reviewer 2: @natebower Added intro sentence.

  "_primary_term": 1
}
```
{% include copy-curl.html %}

</details>

---

## Search monitors

> Query and retrieve information about existing monitors based on specific criteria, such as the monitor name, using the following request.

Reviewer 2: @natebower Added intro sentence.

  "failed": []
}
```
{% include copy-curl.html %}

</details>

---

## Create destination

> Define and configure various destinations for receiving alert notifications using the following request. These destinations can be of different types, such as Slack, custom webhooks, or email, and are used to specify where and how alerts should be delivered.

Reviewer 2: @natebower Added intro sentence.

  ]
}
```
{% include copy-curl.html %}

</details>

---

## Delete destination

> Remove a specific destination from the alerting system using the following request.

Reviewer 2: @natebower Added intro sentence.

  "_primary_term": 1
}
```
{% include copy-curl.html %}

</details>

---

## Create email account

> Set up a new email account for sending alert notifications using the following request.

Reviewer 2: @natebower Added intro sentence.

  }
}
```
{% include copy-curl.html %}

</details>

## Get email account

> Retrieve the details of a specific email account configured for alerting purposes by using the following request.

Reviewer 2: @natebower Added intro sentence.

  }
}
```
{% include copy-curl.html %}

</details>

## Delete email account

> Remove an existing email account configuration from the alerting system using the following request.

Reviewer 2: @natebower Added intro sentence.

  "_primary_term" : 2
}
```
{% include copy-curl.html %}

</details>

## Search email account

> Retrieve information about the configured email accounts used for email-based alerting using the following request.

Reviewer 2: @natebower Added intro sentence.

  }
}
```
{% include copy-curl.html %}

</details>

---

## Create email group

> Define a new group of email recipients for alerts using the following request.

Reviewer 2: @natebower Added intro sentence.

  }
}
```

## Create comment (experimental)

#### Request

```json
POST _plugins/_alerting/comments/<alert-id>
{
  "content": "sample comment"
}
```

#### Example response

```json
{
  "_id": "0U6aBJABVWc3FrmWer9s",
  "_seq_no": 7,
  "_primary_term": 2,
  "comment": {
    "entity_id": "vCZkA5ABWTh3kzuBEL_9",
    "entity_type": "alert",
    "content": "sample comment",
    "created_time": 1718064151148,
    "last_updated_time": null,
    "user": "admin"
  }
}
```

> ## Update comment (experimental)

Reviewer 2: ```suggestion
## Update comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}
```

  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": ".opensearch-alerting-comments-history-2024.06.10-1",
        "_id": "xE5tBJABVWc3FrmWRL5i",
        "_version": 1,
        "_seq_no": 3,
        "_primary_term": 2,
        "_score": 1,
        "_source": {
          "entity_id": "vCZkA5ABWTh3kzuBEL_9",
          "entity_type": "alert",
          "content": "a different sample comment",
          "created_time": 1718061188191,
          "last_updated_time": null,
          "user": "admin"
        }
      },
      {
        "_index": ".opensearch-alerting-comments-history-2024.06.10-1",
        "_id": "0U6aBJABVWc3FrmWer9s",
        "_version": 3,
        "_seq_no": 9,
        "_primary_term": 3,
        "_score": 1,
        "_source": {
          "entity_id": "vCZkA5ABWTh3kzuBEL_9",
          "entity_type": "alert",
          "content": "sample updated comment",
          "created_time": 1718064151148,
          "last_updated_time": 1718064745485,
          "user": "admin"
        }
      }
    ]
  }
}
```

> ## Delete comment (experimental)

Reviewer 2: ```suggestion
## Delete comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}
```

  }
}
```
{% include copy-curl.html %}

</details>

## Get email group

> Retrieve the details of a specific email group destination using the following request, passing the ID of the email group you want to fetch.

Reviewer 2: @natebower Added intro sentence.

  }
}
```
{% include copy-curl.html %}

</details>

## Delete email group

> Remove an existing email group from the list of destinations for alerts using the following request.

Reviewer 2: @natebower Added intro sentence.

  "_primary_term" : 2
}
```
{% include copy-curl.html %}

</details>

## Search email group

> Query and retrieve information about existing email groups used for alerting purposes, enabling you to filter and sort the results based on various criteria. An example is shown in the following request.

Reviewer 2: @natebower Added intro sentence.

    Select to expand example response
  </summary>
  {: .text-delta}
  
#### Example response
```json
{
  "_id": "0U6aBJABVWc3FrmWer9s",
  "_seq_no": 7,
  "_primary_term": 2,
  "comment": {
    "entity_id": "vCZkA5ABWTh3kzuBEL_9",
    "entity_type": "alert",
    "content": "sample comment",
    "created_time": 1718064151148,
    "last_updated_time": null,
    "user": "admin"
  }
}
```
{% include copy-curl.html %}

</details>

## Update comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}

#### Example request

```json
PUT _plugins/_alerting/comments/<comment-id>
{
  "content": "sample updated comment"
}
```
{% include copy-curl.html %}

<details markdown="block">
  <summary>
    Select to expand example response
  </summary>
  {: .text-delta}
  
#### Example response
```json
{
  "_id": "0U6aBJABVWc3FrmWer9s",
  "_seq_no": 8,
  "_primary_term": 3,
  "comment": {
    "entity_id": "vCZkA5ABWTh3kzuBEL_9",
    "entity_type": "alert",
    "content": "sample updated comment",
    "created_time": 1718064151148,
    "last_updated_time": 1718064745485,
    "user": "admin"
  }
}
```
{% include copy-curl.html %}

</details>

> ## Search comments

Reviewer: Should "comment" be singular (as in the preceding headings)?

  }
}
```
{% include copy-curl.html %}

</details>

## Create comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}

> Add comments to a specific alert, providing additional context or notes related to that alert using the following request.

Reviewer 2: @natebower Added intro sentence.

> Add comments to a specific alert, providing additional context or notes related to that alert using the following request.

Reviewer: ```suggestion
Add comments to a specific alert, providing additional context or notes related to that alert, using the following request.
```

#### Example request
```json
POST _plugins/_alerting/comments/<alert-id>
{
  "content": "sample comment"
}
```
{% include copy-curl.html %}

<details markdown="block">
  <summary>
    Select to expand example response
  </summary>
  {: .text-delta}
  
#### Example response
```json
{
  "_id": "0U6aBJABVWc3FrmWer9s",
  "_seq_no": 7,
  "_primary_term": 2,
  "comment": {
    "entity_id": "vCZkA5ABWTh3kzuBEL_9",
    "entity_type": "alert",
    "content": "sample comment",
    "created_time": 1718064151148,
    "last_updated_time": null,
    "user": "admin"
  }
}
```
{% include copy-curl.html %}

</details>

## Update comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}

> Modify the content of a previously added comment associated with an alert using the following request.

Reviewer 2: @natebower Added intro sentence.

#### Example request

```json
PUT _plugins/_alerting/comments/<comment-id>
{
  "content": "sample updated comment"
}
```
{% include copy-curl.html %}

<details markdown="block">
  <summary>
    Select to expand example response
  </summary>
  {: .text-delta}
  
#### Example response
```json
{
  "_id": "0U6aBJABVWc3FrmWer9s",
  "_seq_no": 8,
  "_primary_term": 3,
  "comment": {
    "entity_id": "vCZkA5ABWTh3kzuBEL_9",
    "entity_type": "alert",
    "content": "sample updated comment",
    "created_time": 1718064151148,
    "last_updated_time": 1718064745485,
    "user": "admin"
  }
}
```
{% include copy-curl.html %}

</details>

## Search comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}

> Query and retrieve existing comments associated with alerts using the following request.

Reviewer 2: @natebower Added intro sentence.

    "hits": [
      {
        "_index": ".opensearch-alerting-comments-history-2024.06.10-1",
        "_id": "xE5tBJABVWc3FrmWRL5i",
        "_version": 1,
        "_seq_no": 3,
        "_primary_term": 2,
        "_score": 1,
        "_source": {
          "entity_id": "vCZkA5ABWTh3kzuBEL_9",
          "entity_type": "alert",
          "content": "a different sample comment",
          "created_time": 1718061188191,
          "last_updated_time": null,
          "user": "admin"
        }
      },
      {
        "_index": ".opensearch-alerting-comments-history-2024.06.10-1",
        "_id": "0U6aBJABVWc3FrmWer9s",
        "_version": 3,
        "_seq_no": 9,
        "_primary_term": 3,
        "_score": 1,
        "_source": {
          "entity_id": "vCZkA5ABWTh3kzuBEL_9",
          "entity_type": "alert",
          "content": "sample updated comment",
          "created_time": 1718064151148,
          "last_updated_time": 1718064745485,
          "user": "admin"
        }
      }
    ]
  }
}
```
{% include copy-curl.html %}

</details>

## Delete comment
This is an experimental feature and is not recommended for use in a production environment.   
{: .warning}

> Remove a specific comment associated with an alert using the following request.

Reviewer 2: @natebower Added intro sentence.
